Mathlib/GroupTheory/SpecificGroups/Cyclic.lean

```diff
@@ -680,3 +680,90 @@ lemma not_isCyclic_iff_exponent_eq_prime [Group α] {p : ℕ} (hp : p.Prime)
 
 end Exponent
+
+section ZMod
+
+open Subgroup AddSubgroup
+
+variable {G H : Type*}
+
+/-- The isomorphism between `(Multiplicative G) / ⟨.ofAdd g⟩` and `Multiplicative (G / ⟨g⟩)`. -/
+def multiplicativeQuotientZpowersZmultiplesMulEquiv [AddCommGroup G] (g : G) :
+    Multiplicative G ⧸ zpowers (Multiplicative.ofAdd g) ≃* Multiplicative (G ⧸ zmultiples g) :=
+  MulEquiv.refl _
+
+lemma multiplicativeQuotientZpowersZmultiplesMulEquiv_mk_eq [AddCommGroup G] (g : G)
+    (h : Multiplicative G) :
+    multiplicativeQuotientZpowersZmultiplesMulEquiv g (QuotientGroup.mk h) =
+    .ofAdd (QuotientAddGroup.mk h) := rfl
+
+lemma multiplicativeQuotientZpowersZmultiplesMulEquiv_symm_mk_eq [AddCommGroup G] (g : G)
+    (h : G) :
+    (multiplicativeQuotientZpowersZmultiplesMulEquiv g).symm (QuotientAddGroup.mk h) =
+    QuotientGroup.mk (.ofAdd h) := rfl
+
+/-- The kernel of `zpowersHom G g` is equal to the subgroup generated by `orderOf g`. -/
+theorem zpowersHom_ker_eq [Group G] (g : G) :
+    (zpowersHom G g).ker = zpowers (Multiplicative.ofAdd ↑(orderOf g)) := by
+  ext n
+  rw [MonoidHom.mem_ker, mem_zpowers_iff, zpowersHom_apply, ← orderOf_dvd_iff_zpow_eq_one]
+  let m : ℤ := n
+  change _ ∣ m ↔ ∃ k, k * (orderOf g : ℤ) = m
+  conv in _ * _ = _ => rw [mul_comm, eq_comm]
+
+/-- The kernel of `zmultiplesHom G g` is equal to the additive subgroup generated by
+    `addOrderOf g`. -/
+theorem zmultiplesHom_ker_eq [AddGroup G] (g : G) :
+    (zmultiplesHom G g).ker = zmultiples ↑(addOrderOf g) := by
+  ext n
+  rw [AddMonoidHom.mem_ker, mem_zmultiples_iff, zmultiplesHom_apply,
+    ← addOrderOf_dvd_iff_zsmul_eq_zero]
+  let m : ℤ := n
+  change _ ∣ m ↔ ∃ k, k * (addOrderOf g : ℤ) = m
+  conv in _ * _ = _ => rw [mul_comm, eq_comm]
+
+/-- The isomorphism from `ZMod n` to any cyclic additive group of `Nat.card` equal to `n`. -/
+noncomputable def zmodAddCyclicAddEquiv [AddGroup G] (h : IsAddCyclic G) :
+    ZMod (Nat.card G) ≃+ G := by
+  let n := Nat.card G
+  let ⟨g, surj⟩ := Classical.indefiniteDescription _ h.exists_generator
+  have kereq : ((zmultiplesHom G) g).ker = zmultiples ↑(Nat.card G) := by
+    rw [zmultiplesHom_ker_eq]
+    congr
+    rw [← Nat.card_zmultiples]
+    exact Nat.card_congr (Equiv.subtypeUnivEquiv surj)
+  exact (Int.quotientZmultiplesNatEquivZMod n)
+    |>.symm.trans <| QuotientAddGroup.quotientAddEquivOfEq kereq
+    |>.symm.trans <| QuotientAddGroup.quotientKerEquivOfSurjective (zmultiplesHom G g) surj
+
+/-- The isomorphism from `Multiplicative (ZMod n)` to any cyclic group of `Nat.card` equal to `n`.
+-/
+noncomputable def zmodCyclicMulEquiv [Group G] (h : IsCyclic G) :
+    Multiplicative (ZMod (Nat.card G)) ≃* G :=
+  AddEquiv.toMultiplicative <| zmodAddCyclicAddEquiv <| isAddCyclic_additive_iff.2 h
+
+/-- Two cyclic groups of the same cardinality are isomorphic. -/
+noncomputable def mulEquivOfCyclicCardEq [Group G] [Group H] (hG : IsCyclic G)
+    (hH : IsCyclic H) (hcard : Nat.card G = Nat.card H) : G ≃* H := hcard ▸
```

Comment:
`IsCyclic` and `IsAddCyclic` are type classes, so the arguments  can be changed to `[hG : IsCyclic G] [hH : IsCyclic H]` (and similarly for the additive version).

In the proof of `mulEquivOfPrimeCardEq` below, you can use `(hG := isCyclic_of_prime_card hG)` to specify the instance or, use `have hGcyc := isCyclic_of_prime_card hG` (and similarly for `H`) before the `apply` and then leave them  out.